### DISCUSSION TITLE
Is there a way to build and maintain multi language site in jupyter book using only one repo?

### DISCUSSION THREAD

**Discussion by SDSTony** [executablebooks/meta#220] [Q&A] [answered]
I would like to host multi language versions (Korean, English) of my [jupyter book website](https://pseudo-lab.github.io/Tutorial-Book/). 

Currently, the only way that I can think of is to create a repository for each versions of website. (ex. Jupyter-Book-Kor, Jupyter-Book-Eng)

However, this would be inconvenient in maintenance wise since I have to keep track of two different repos where the only difference is the language of the content files and table of contents. 

I wonder if there is a way to build and maintain multi language versions of jupyter book website in one repo. 

Thanks.

**Comment by chrisjsewell** [comment, marked as answer]
I'm not sure if it is possible in one repo, and I've tried with jupyter-book.
But with a general sphinx site, I believe the recommended way is detailed here: https://www.sphinx-doc.org/en/master/usage/advanced/intl.html, using [transifex](https://docs.transifex.com/integrations/sphinx-doc)

Here is an example from a project I am involved with, to translate to chinese: https://github.com/unkcpz/aiida-l10n-zh_CN

**Reply by SDSTony** [reply]
Thanks for you help! I'll refer to the attached site along with [Custom Sphinx Configuration guide of Jupyter Book](https://jupyterbook.org/advanced/sphinx.html) to see if there is a way to pull this off. 

For the time being, I've decided to go with creating multiple repos. 

By the way, thanks for open sourcing Jupyter Book. I absolutely love it.

**Reply by chrisjsewell** [reply]
thanks! certainly report back if you see any opportunities to improve multi-language support 😄

**Reply by choldgraf** [reply]
indeed please do let us know how this goes, it would be really great to document any patterns that are useful for others to follow

**Comment by Daltz333** [comment]
I'm not sure of the integration between sphinx and executablebooks. But https://github.com/wpilibsuite/frc-docs-translations is a 2-repo system that could probably be combined into one (just a separate github actions workflow). It's documented a fair bit and I'd be happy to answer any questions

**Comment by anjackson** [comment]
I've just been attempting to set up a multilingual JupyterBook site, and got so far before hitting a problem that I can't seem to resolve.  So, I'm going to document what I did here, in case someone else can take it further.

The project indicated above using things like `sphinx-build -b gettext docs/source locale` to start the translation process, but that doesn't work as-is because we're using the JupyterBook `jb build .` wrapper (note that I'm assuming the source is in the current working directory). Eventually, I managed to work out that the correct incantation is:

    jb build --builder custom --custom-builder gettext --path-output _gettext .

Note that I'm using a different output path, because if you don't you get this error:

    This environment is incompatible with the selected builder, please choose another doctree directory.

To get this to work at all, I needed these settings:

```
sphinx:
  config:
    locale_dirs: ['locale/']
    gettext_compact: false
    nb_render_priority:
      gettext: {}
```
However, the attempt to build the translation files still fails, with this error:

```
  File "/Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx/util/osutil.py", line 55, in canon_path
    return nativepath.replace(path.sep, SEP)
AttributeError: 'NoneType' object has no attribute 'replace'
```

Which seems to be similar to [this old Sphinx issue related to image blocks](https://github.com/sphinx-doc/sphinx/issues/3093).  I can hack Sphinx sidestep this error, but as [that](https://github.com/doctorlard/sphinx/commit/b6cd676d888c55b1e9cf4e13589651497c7368f0) was not part of [the final fix for that issue](https://github.com/sphinx-doc/sphinx/pull/3165), I'm not sure how to proceed.

**Reply by Daltz333** [reply]
Can you share the entire stacktrace? I want to see what extension or configuration file calls that? I am not familiar with the juypterbook side of things, but I do know that Juypterbook very intimitely integrates into Sphinx. I know it's possible to go from Sphinx -> Juypter book, but is Juypterbook backwards compatible from Juypterbook -> Sphinx?

@choldgraf ?

**Reply by anjackson** [reply]
Thanks! I _think_ the intent is that with JupyterBook you get all the benefits of using Sphinx plus some additional Jupyter functionality on top. 

Here's the trace:

```
# Sphinx version: 3.5.4
# Python version: 3.9.5 (CPython)
# Docutils version: 0.16 release
# Jinja2 version: 3.0.1
# Last messages:
#   writing message catalogs... [ 59%] _gettext/_build/jupyter_execute/jupyter_execute/how-ukwa-works/arch-plots/render-arch-plots
#   writing message catalogs... [ 60%] _gettext/_build/jupyter_execute/jupyter_execute/how-ukwa-works/arch-plots/render-stacks
#   writing message catalogs... [ 60%] _gettext/_build/jupyter_execute/jupyter_execute/how-ukwa-works/maps
#   writing message catalogs... [ 61%] how-ukwa-works/_index
#   writing message catalogs... [ 61%] how-ukwa-works/arch-plots/render-arch-plots
#   writing message catalogs... [ 62%] how-ukwa-works/arch-plots/render-stacks
#   writing message catalogs... [ 62%] how-ukwa-works/arch/access
#   writing message catalogs... [ 63%] how-ukwa-works/arch/ingest
#   writing message catalogs... [ 63%] how-ukwa-works/arch/manage
#   writing message catalogs... [ 64%] how-ukwa-works/arch/overview
# Loaded extensions:
#   sphinx.ext.mathjax (3.5.4) from /Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx/ext/mathjax.py
#   sphinxcontrib.applehelp (1.0.2) from /Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinxcontrib/applehelp/__init__.py
#   sphinxcontrib.devhelp (1.0.2) from /Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinxcontrib/devhelp/__init__.py
#   sphinxcontrib.htmlhelp (2.0.0) from /Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinxcontrib/htmlhelp/__init__.py
#   sphinxcontrib.serializinghtml (1.1.5) from /Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinxcontrib/serializinghtml/__init__.py
#   sphinxcontrib.qthelp (1.0.3) from /Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinxcontrib/qthelp/__init__.py
#   alabaster (0.7.12) from /Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/alabaster/__init__.py
#   sphinx_togglebutton (0.2.3) from /Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx_togglebutton/__init__.py
#   sphinx_copybutton (0.3.1) from /Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx_copybutton/__init__.py
#   myst_nb (0.12.3) from /Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/myst_nb/__init__.py
#   jupyter_book (0.11.1) from /Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/jupyter_book/__init__.py
#   sphinx_thebe (0.0.8) from /Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx_thebe/__init__.py
#   sphinx_comments (0.0.3) from /Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx_comments/__init__.py
#   sphinx_external_toc (0.2.1) from /Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx_external_toc/__init__.py
#   sphinx.ext.intersphinx (3.5.4) from /Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx/ext/intersphinx.py
#   sphinx_panels (0.5.2) from /Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx_panels/__init__.py
#   sphinx_book_theme (unknown version) from /Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx_book_theme/__init__.py
#   jupyterbook_latex (unknown version) from /Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/jupyterbook_latex/__init__.py
Traceback (most recent call last):
  File "/Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/jupyter_book/sphinx.py", line 150, in build_sphinx
    app.build(force_all, filenames)
  File "/Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx/application.py", line 352, in build
    self.builder.build_update()
  File "/Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 296, in build_update
    self.build(to_build,
  File "/Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx/builders/gettext.py", line 284, in build
    super().build(docnames, summary, method)
  File "/Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 363, in build
    self.finish()
  File "/Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx/builders/gettext.py", line 307, in finish
    content = GettextRenderer(outdir=self.outdir).render('message.pot_t', context)
  File "/Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx/builders/gettext.py", line 130, in render
    return super().render(filename, context)
  File "/Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx/util/template.py", line 34, in render
    return self.env.get_template(template_name).render(context)
  File "/Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/jinja2/environment.py", line 1304, in render
    self.environment.handle_exception()
  File "/Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/jinja2/environment.py", line 925, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "/Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx/templates/gettext/message.pot_t", line 21, in top-level template code
    #: {{ relpath(source) }}:{{ line }}
  File "/Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/jinja2/sandbox.py", line 393, in call
    return __context.call(__obj, *args, **kwargs)
  File "/Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx/builders/gettext.py", line 127, in _relpath
    return canon_path(relpath(s, self.outdir))
  File "/Users/andy/Documents/workspace/ukwa-documentation/venv/lib/python3.9/site-packages/sphinx/util/osutil.py", line 58, in canon_path
    return nativepath.replace(path.sep, SEP)
AttributeError: 'NoneType' object has no attribute 'replace'
```

**Reply by anjackson** [reply]
Ah, I think JupyterBook is doing something that means the `message.locations` are not being set correctly (with `source` being `None`?).  If I switch off `display_location` (using `gettext_location: false` in `config.yml`) it runs fine.

**Reply by choldgraf** [reply]
Jupyter Book shouldn't be doing much (if any) magic under the hood, everything just becomes a Sphinx build, so this should definitely be possible but it might take some fiddling :-/

**Reply by anjackson** [reply]
So far the fiddling looks like (snippet of `_config.yml`):

```
sphinx:
  config:
    locale_dirs: ['locales/']
    gettext_location: false
    nb_render_priority:
      gettext: {}
```

But I _think_ it's working now.

**Reply by gsuarezr** [reply]
Could you please link to your jupyter book? I want to do the same and want to take a look at the end result

**Reply by anjackson** [reply]
Sorry @mcditoos I don't have a complete example to share yet. I've just done enough to determine that it should work, but it'll be a while before I have time to do more.

**Reply by hammerdirt-analyst** [reply]
Hi,

I am totally interested in this  and willing to put the time in.  We are delivering to the client two seperate books one in English and one in German. However we would prefer one. Actually we need three or four languages. In Switzerland we have three official languages.

The book is here: https://hammerdirt-analyst.github.io/IQAASL-End-0f-Sampling-2021/titlepage.html

The repo is being refactored right now so no judgments please. We have spent mulitple years in the field building this as we go (the release of Jupyter books really helped us out).  I would need some time to get up to speed with the Sphinx-Jupyter interface but would really like to help out on this.

**Reply by hammerdirt-analyst** [reply]
We did a multilanguage report in English and German in one repo. Our use case is a little special. There is a note book in each language. We run jupyter book build ./ using different congfig and toc files (one for each language)

https://hammerdirt-analyst.github.io/IQAASL-End-0f-Sampling-2021/titlepage.html

**Reply by whyjz** [reply]
Hi @anjackson and @choldgraf and others, 
Just wanted to write down my observations and potential cause about the `AttributeError: 'NoneType' object has no attribute 'replace'` for future reference. This happens when `_toc.yml` uses at least one `- caption` (together with `parts`) to structure the content. I suppose it also happens when there are other user-defined strings in the TOC but have not tested them through.

With `gettext_location` set to `True`, `jb build --builder custom --custom-builder gettext --path-output _gettext .` runs fine until it hits the root page (say `index.md`). All the TOC caption strings are parsed and stored in the translation template `index.pot`, but it is not possible for `gettext` to find where these strings are in `index.md` because they are actually from `_toc.yml`. To avoid this error, `gettext_location` must be set to `False` to skip the problematic step.

**Comment by whyjz** [comment]
@SDSTony Yes, this is what I have been doing for [this tutorial website](https://gmt-tutorials.org/en/index.html) and the [corresponding repository](https://github.com/whyjz/GMT-tutorials). 

Here are specific steps. I assume the book content is in the `docs` folder.
1. Make gettext files (translation templates for a new language to fresh start) 
   ```
   jupyter book build --builder custom --custom-builder gettext docs
   ```
2. Specify a target language, make .po files, and start translating. See sphinx-intl for more details. My `[locale_folder]` is set to `docs/locale`. This should agree with the `locale_dirs` settings defined in `_config.yml`. 

   ```
   sphinx-intl update -p docs/_build/gettext -d [locale_folder] -l [target_language]
   ```

3. Generate HTML files in the source language.
   ```
   jupyter book build docs
   ```

4. Generate HTML files in the target language. Note that Jupyter Book's CLI does not allow overriding settings defined in `_config.yml` (i.e., the -D option in `sphinx-build`), so we have to generate `config.py` and use `sphinx-build` to do the step.
   ```
   jupyter book config sphinx docs
   sphinx-build -b html -D language=[target_language] docs [path_for_htmls]
   ```

5. Your native-language HTMLs are now in `docs/_build/html/` and the translated HTMLs are available in [path_for_htmls]. 

Here are my relevant `_config.yml` settings for your reference:

```yml
sphinx:
  config:
    language: zh_TW              # defining the source language 
    html_search_language: ja     # ja is better at parsing zh_TW strings for search box
    # i18n related
    nb_render_priority:
      gettext: {}                # see the discussion above
    locale_dirs: ['locale/']     # my _config.yml is in docs/, so this points to docs/locale
    gettext_compact: false
    gettext_additional_targets: ['literal-block']
    gettext_location: false      # see the discussion above
```

**Reply by fccoelho** [reply]
This should be added to the documentation.

**Reply by hammerdirt-analyst** [reply]
Okay, I am waiting on some translations. When I get them I will try to implement (there is three languages and 20 +chapters/workbooks). Maybe I can help with the documentation after the experiene?
